File: solutions/fortran/rational-numbers/1/rational_numbers.f90
module rational_numbers
  implicit none
contains

  ! Helper: greatest common divisor
  integer function gcd(a, b)
    integer, intent(in) :: a, b
    integer :: x, y, t
    x = abs(a)
    y = abs(b)
    do while (y /= 0)
       t = y
       y = mod(x, y)
       x = t
    end do
    gcd = x
  end function gcd

  ! Reduce rational number to lowest terms with positive denominator
  function reduce(r) result(res)
    integer, dimension(2), intent(in) :: r
    integer, dimension(2) :: res
    integer :: common
    common = gcd(r(1), r(2))
    res(1) = r(1) / common
    res(2) = r(2) / common
    if (res(2) < 0) then
       res(1) = -res(1)
       res(2) = -res(2)
    end if
  end function reduce

  ! Addition
  function add(r1, r2) result(res)
    integer, dimension(2), intent(in) :: r1, r2
    integer, dimension(2) :: res
    integer :: num, den
    num = r1(1)*r2(2) + r2(1)*r1(2)
    den = r1(2)*r2(2)
    res = reduce([num, den])
  end function add

  ! Subtraction
  function sub(r1, r2) result(res)
    integer, dimension(2), intent(in) :: r1, r2
    integer, dimension(2) :: res
    integer :: num, den
    num = r1(1)*r2(2) - r2(1)*r1(2)
    den = r1(2)*r2(2)
    res = reduce([num, den])
  end function sub

  ! Multiplication
  function mul(r1, r2) result(res)
    integer, dimension(2), intent(in) :: r1, r2
    integer, dimension(2) :: res
    integer :: num, den
    num = r1(1)*r2(1)
    den = r1(2)*r2(2)
    res = reduce([num, den])
  end function mul

  ! Division
  function div(r1, r2) result(res)
    integer, dimension(2), intent(in) :: r1, r2
    integer, dimension(2) :: res
    integer :: num, den
    num = r1(1)*r2(2)
    den = r2(1)*r1(2)   ! r2(1) assumed non-zero
    res = reduce([num, den])
  end function div

  ! Absolute value
  function rational_abs(r1) result(res)
    integer, dimension(2), intent(in) :: r1
    integer, dimension(2) :: res
    res = reduce([abs(r1(1)), abs(r1(2))])
  end function rational_abs

  ! Exponentiation to an integer power
  function rational_to_pow(r1, ex) result(res)
    integer, dimension(2), intent(in) :: r1
    integer, intent(in) :: ex
    integer, dimension(2) :: res
    integer :: num, den, p

    if (ex >= 0) then
       num = r1(1)**ex
       den = r1(2)**ex
    else
       p = -ex
       num = r1(2)**p
       den = r1(1)**p
    end if
    res = reduce([num, den])
  end function rational_to_pow

  ! Exponentiation of a real number to a rational power
  function real_to_rational_pow(ex, r1) result(res)
    real, intent(in) :: ex
    integer, dimension(2), intent(in) :: r1
    real :: res
    res = ex ** (real(r1(1)) / real(r1(2)))
  end function real_to_rational_pow

end module rational_numbers

File: solutions/c/rational-numbers/1/rational_numbers.c
#include "rational_numbers.h"
#include <stdlib.h>
#include <math.h>

// Helper: greatest common divisor
static int gcd(int a, int b) {
    a = abs(a);
    b = abs(b);
    
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

// Helper: create reduced rational number
static rational_t create(int num, int den) {
    rational_t r = {num, den};
    
    if (den == 0) {
        return r;  // As per note, denominator can be 0
    }
    
    // Move sign to numerator
    if (den < 0) {
        r.numerator = -num;
        r.denominator = -den;
    }
    
    // Reduce to lowest terms
    int divisor = gcd(r.numerator, r.denominator);
    if (divisor != 0) {
        r.numerator /= divisor;
        r.denominator /= divisor;
    }
    
    return r;
}

rational_t add(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.denominator + r2.numerator * r1.denominator;
    int den = r1.denominator * r2.denominator;
    return create(num, den);
}

rational_t subtract(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.denominator - r2.numerator * r1.denominator;
    int den = r1.denominator * r2.denominator;
    return create(num, den);
}

rational_t multiply(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.numerator;
    int den = r1.denominator * r2.denominator;
    return create(num, den);
}

rational_t divide(rational_t r1, rational_t r2) {
    int num = r1.numerator * r2.denominator;
    int den = r1.denominator * r2.numerator;
    return create(num, den);
}

rational_t absolute(rational_t r) {
    int num = abs(r.numerator);
    int den = abs(r.denominator);
    return create(num, den);
}

rational_t exp_rational(rational_t r, int n) {
    if (n == 0) {
        return create(1, 1);
    }
    
    rational_t result = create(1, 1);
    rational_t base = r;
    int power = abs(n);
    
    while (power > 0) {
        if (power & 1) {
            result = multiply(result, base);
        }
        base = multiply(base, base);
        power >>= 1;
    }
    
    if (n < 0) {
        return divide(create(1, 1), result);
    }
    
    return result;
}

float exp_real(int x, rational_t r) {
    // x^(r.numerator/r.denominator) = (x^r.numerator)^(1/r.denominator)
    double base_pow_num = pow(x, r.numerator);
    return (float)pow(base_pow_num, 1.0 / r.denominator);
}

rational_t reduce(rational_t r) {
    return create(r.numerator, r.denominator);
}